docs/concepts/what-is-a-workflow-definition.md

```diff
@@ -11,13 +11,97 @@ A Workflow Definition is the code that defines the constraints of a Workflow Exe
+
+The execution semantics of a Workflow Execution include the re-execution of a Workflow Function.
+The use of Workflow APIs in the function generate the Commands.
+Commands tell the Cluster what Events to create and add to the Workflow Execution's Event History.
+Whenever a Workflow Function is executed, the Commands that are emitted are compared with the existing Event History.
+If a corresponding Event already exists within the Event History that maps to the generation of that Command in the same sequence, and some specific metadata of that Command matches with some specific metadata of the Event, then that Command is ignored.
+
+For example, using an SDKs "Execute Activity" API generates the [ScheduleActivityTask](/docs/concepts/what-is-a-command#scheduleactivitytask) Command.
+When this API is called upon re-execution, that Command is compared with the Event that is in the same location within the sequence.
+The Event in the sequence must be an [ActivityTaskScheduled](/docs/concepts/what-is-an-event/#activitytaskscheduled) Event, where the Activity Name and the Task Queue name are the same as what is in the Command.
+
+If a Command is generated that doesn't match what it needs to in the existing Event History, then the Workflow Execution will return a "non-deterministic" error.
+
+The following are the two reasons why a Command might be generated out of sequence or the wrong Command might be generated altogether:
+
+1. Code changes are made to a Workflow Definition that is in use by a running Workflow Execution.
+2. There is intrinsic non-deterministic logic (i.e. inline random branching logic).
+
+#### Non-deterministic code changes
+
+The Workflow Definition can change in very limited ways once there is a Workflow Execution depending on it.
+To alleviate non-deterministic issues that arise from code changes we recommend using [Workflow Versioning](#what-is-workflow-versioning).
+
+For example, let's say we have a Workflow Definition that defines the following sequence:
+
+1. Start and wait on a Timer/sleep
+2. Spawn and wait on an Activity Execution
+3. Complete
+
+We start a Worker and spawn a Workflow Execution that uses that Workflow Definition.
+The Worker would emit the [StartTimer](/docs/concepts/what-is-a-command/#starttimer) Command and the Workflow Execution would become suspended.
+
+Before the Timer is up, we change the Workflow Definition the following sequence:
+
+1. Spawn and wait on an Activity Execcution
+2. Start and wait on a timer/sleep
+3. Complete
+
+When the Timer fires, the next Workflow Task will cause the Workflow Function to re-execute.
+The first Command it sees would be be ScheduleActivityTask, which wouldn't match up to the expected ActivityTaskScheduled Event.
+
+The Workflow Execution would fail, and return the non-determinism error.
+
+There following are examples of some minor changes that would not take effect when re-executing a History which already contain the Events:
```

Comment:
This is true but I think the more important thing is to say that they won't cause command mismatches / nondeterminism errors.